app/src/test/kotlin/compiler/parser/analysis/GrammarAnalysisTest.kt

```diff
@@ -1,5 +1,1171 @@
+    //      Follow(start) = { }
+    //
+    //      First(A) = { A, B }
+    //      Follow(A) = { }
+    //
+    //      First(A) = { B }
+    //      Follow(A) = { }
+    //
+
+    private val symA = "A"
+    private val symB = "B"
+
+    private val dfaStatesA = listOf("startState", "accState").map { it to TestDfaState(it) }.toMap()
+    private val dfaDescriptionA = mapOf(
+        Pair(dfaStatesA["startState"]!!, symA) to dfaStatesA["accState"]!!
+    )
+
+    private val dfaStatesB = listOf("startState", "accState").map { it to TestDfaState(it) }.toMap()
+    private val dfaDescriptionB = mapOf(
+        Pair(dfaStatesB["startState"]!!, symB) to dfaStatesB["accState"]!!
+    )
+
+    private val dfaA = TestDfa(dfaStatesA["startState"]!!, dfaDescriptionA)
+    private val dfaB = TestDfa(dfaStatesB["startState"]!!, dfaDescriptionB)
+
+    private val linearNullableGrammar: AutomatonGrammar<String> = AutomatonGrammar(
+        start,
+        mapOf(
+            start to dfaA,
+            symA to dfaB,
+            symB to getTrivialDfa()
+        )
+    )
+
+    @Test
+    fun `test nullable for linear grammar with all nullable`() {
+        val expectedNullable = setOf(start, symA, symB)
+        val actualNullable = GrammarAnalysis<GrammarSymbol>().computeNullable(linearNullableGrammar)
+
+        assertEquals(expectedNullable, actualNullable)
+    }
+
+    @Test
+    fun `test first for linear grammar with all nullable`() {
+        val nullable = setOf(start, symA, symB)
+
+        val expectedFirst = mapOf(
+            start to setOf(start, symA, symB),
+            symA to setOf(symA, symB),
+            symB to setOf(symB)
+        )
+        val actualFirst = GrammarAnalysis<GrammarSymbol>().computeFirst(linearNullableGrammar, nullable)
+
+        assertEquals(expectedFirst, actualFirst)
+    }
+
+    @Test
+    fun `test follow for linear grammar with all nullable`() {
+        val nullable = setOf(start, symA, symB)
+        val first = mapOf(
+            start to setOf(start, symA, symB),
+            symA to setOf(symA, symB),
+            symB to setOf(symB)
+        )
+
+        val expectedFollow = mapOf<GrammarSymbol, Set<GrammarSymbol>>(
+            start to setOf(),
+            symA to setOf(),
+            symB to setOf(),
+        )
+        val actualFollow = GrammarAnalysis<GrammarSymbol>().computeFollow(linearNullableGrammar, nullable, first)
+
+        assertEquals(expectedFollow, actualFollow)
+    }
+
+    // SIMPLE GRAMMAR WITH NOT ALL NULLABLE
+    // Grammar:
+    //   Sigma = { start, c, D }
+    //   Productions = {
+    //      start --> cD (dfaCD)
+    //      D --> eps (trivial dfa)
+    //   }
+    //
+    //      Nullable = { D }
+    //
+    //      First(start) = { start, c }
+    //      Follow(start) = { }
+    //
+    //      First(c) = { c }
+    //      Follow(c) = { D }
+    //
+    //      First(D) = { D }
+    //      Follow(D) = { }
+    //
+
+    private val symc = "c"
```

Comment:
I don't like hardcoded because it makes it easier to do a typo. And I don't like very short names of variables, thus, symc to make it at least a little bit meaningful.